jetstream/analysis.py

```diff
@@ -439,34 +473,57 @@ def run(self, current_date: datetime, dry_run: bool = False) -> None:
-                    "Not calculating statistics %s (%s); dry run",
-                    self.config.experiment.normandy_slug,
-                    period.value,
-                )
-                results.append(metrics_table)
-                continue
+            analysis_bases = []
 
-            metrics_data = table_to_dataframe(metrics_table)
+            for m in self.config.metrics[period]:
+                for analysis_basis in m.metric.analysis_bases:
+                    analysis_bases.append(analysis_basis)
 
-            segment_results = []
+            analysis_bases = list(set(analysis_bases))
 
-            segment_labels = ["all"] + [s.name for s in self.config.experiment.segments]
-            for segment in segment_labels:
-                segment_data = self.subset_to_segment(segment, metrics_data)
-                for m in self.config.metrics[period]:
-                    segment_results += self.calculate_statistics(
-                        m,
-                        segment_data,
-                        segment,
-                    ).to_dict()["data"]
+            if len(analysis_bases) == 0:
+                continue
 
-                segment_results += self.counts(segment_data, segment).to_dict()["data"]
+            for analysis_basis in analysis_bases:
+                metrics_table = self.calculate_metrics(
+                    exp, time_limits, period, analysis_basis, dry_run
+                )
 
-            results.append(self.save_statistics(period, segment_results, metrics_table))
+                if dry_run:
+                    results.append(metrics_table)
+                else:
+                    metrics_dataframe = table_to_dataframe(metrics_table)
+
+                if dry_run:
+                    logger.info(
+                        "Not calculating statistics %s (%s); dry run",
+                        self.config.experiment.normandy_slug,
+                        period.value,
+                    )
+                    continue
+
+                segment_labels = ["all"] + [s.name for s in self.config.experiment.segments]
+                for segment in segment_labels:
+                    segment_data = self.subset_to_segment(segment, metrics_dataframe)
+                    for m in self.config.metrics[period]:
+                        segment_results += self.calculate_statistics(
+                            m,
+                            segment_data,
+                            segment,
+                            analysis_basis,
+                        ).to_dict()["data"]
+
+                        segment_results += self.counts(
+                            segment_data, segment, analysis_basis
+                        ).to_dict()["data"]
+
+        results.append(
+            self.save_statistics(
+                period,
+                segment_results,
+                self._table_name(period.value, len(time_limits.analysis_windows)),
+            )
+        )
```

Comment:
I guess we can just compute segments for each analysis basis separately and then store the results to the statistics table all together.